replicator/src/main/scala/com/evolutiongaming/kafka/journal/replicator/Batch.scala

```diff
@@ -14,139 +24,83 @@ private[journal] object Batch {
+      case (state, _) if state.purge.nonEmpty =>
+        state
+
+      case (state, ActionRecord(purge: Action.Purge, po)) =>
+        state.copy(
+          purge = Purge(po.offset, purge.origin, purge.version).some,
+        )
+
+      case (state, ActionRecord(delete: Action.Delete, po)) =>
+        val delete_ = state.delete match {
+          case Some(younger) =>
+            // take `origin` and `version` from "older" entity, if it has them
+            val origin  = delete.origin.orElse(younger.origin)
+            val version = delete.version.orElse(younger.version)
+            // make `Delete` action with largest `seqNr` and largest `offset`
+            if (younger.to < delete.to) Delete(po.offset, delete.to, origin, version)
+            else younger.copy(origin = origin, version = version)
+
+          case None =>
+            Delete(po.offset, delete.to, delete.origin, delete.version)
         }
-
-        def deleteOf(to: DeleteTo, origin: Option[Origin], version: Option[Version]): Delete = {
-          Delete(offset, to, origin, version)
+        state.copy(
+          delete = delete_.some,
+        )
+
+      case (state, ActionRecord(append: Action.Append, po)) if state.delete.forall(_.to.value <= append.range.to) =>
+        val appends = state.appends match {
+          case Some(appends) =>
+            appends.copy(records = ActionRecord(append, po) :: appends.records)
+          case None =>
+            Appends(po.offset, NonEmptyList.of(ActionRecord(append, po)))
         }
+        state.copy(
+          appends = appends.some,
+        )
 
-        def purgeOf(origin: Option[Origin], version: Option[Version]): Purge = {
-          Purge(offset, origin, version)
-        }
-
-        def actionRecord[A <: Action](a: A): ActionRecord[A] = {
-          record.copy(action = a)
-        }
+      case (state, _) =>
+        // ignore `Action.Append`, if it would get deleted and ignore `Action.Mark`
+        state
+    }
 
-        def origin: Option[Origin] = {
-          bs.foldRight(none[Origin]) { (b, origin) =>
-            origin orElse {
-              b match {
-                case b: Batch.Delete  => b.origin
-                case _: Batch.Appends => none
-                case b: Batch.Purge   => b.origin
-              }
-            }
-          }
-        }
+    state.batches
+  }
 
-        def version: Option[Version] = {
-          bs.foldRight(none[Version]) { (b, version) =>
-            version orElse {
-              b match {
-                case b: Batch.Delete  => b.version
-                case _: Batch.Appends => none
-                case b: Batch.Purge   => b.version
-              }
-            }
+  private final case class State(
+    purge: Option[Purge]     = None,
+    appends: Option[Appends] = None,
+    delete: Option[Delete]   = None,
+  ) {
+
+    def batches: List[Batch] = {
+      // we can drop first `append`, if `deleteTo` will discard it AND there is at least one more `append`
```

Comment:
I added extra comment in code:
```scala
// we have to preserve one `append` to store latest `seqNr` and populate `expireAfter`
```

the alternative was to introduce dedicated new "action" to update only `seqNr` and `expireAfter` fields in `metajournal`, but that was too invasive (in #676 there was an attempt to do so, but it didn't handle `expireAfter` case)

Reply:
Sorry but I don't get the answer on @rtar question: why only _first_ event dropped but not _all (until `deleteTo`)_ except one to carry over `seqNr` and `expireAfter`?

Reply:
because we drop all of them before during batching, but we keep the `Append` where `range.to == deleteTo.seqNr` - we might add more `append`s after it and if we do add more, then we can drop the first one - all other ones will have `seqNr > deleteTo.seqNr` and we have to store all other `append`s in `journal` table

Reply:
example:
```
append(5)
delete(6)
append(6)
append(7)
```
leads to batch: `appends(6, 7); delete(6)` - we have dropped `append(5)` and we can still drop `append(6)`.

in case if there would not be `append(7)`, we would have to keep `append(6)`